### ISSUE TITLE
Pubspec.yaml constraints resolved to unexpected version

### ISSUE THREAD

**Issue by joeconwaystk** [dart-lang/pub#2446]
`package:aqueduct` has versions on pub.dev `3.2.1`, `3.3.0`, and `3.3.0-b1`. The constraint `^3.0.0` selects `3.3.0-b1`. I would expect it to select `3.3.0`. 

It also selected `3.3.0-b1` when only `3.2.1` and `3.3.0-b1` existed (and `3.3.0` did not). If both `3.3.0-b1` and `3.3.0` can simply be deleted from pub.dev, that is best, however, I'd also like to understand what creates that behavior.

pub 2.7.0 macOS.

```yaml
name: foobar
description: An empty Aqueduct application.
version: 0.0.1

environment:
  sdk: ">=2.0.0 <3.0.0"

dependencies:
  aqueduct: ^3.0.0

dev_dependencies:
  test: ^1.0.0
  aqueduct_test: ^1.0.0
```

**Comment by jakemac53** [comment]
tldr; it chose a version solve of other dependencies which ended up only allowing the prerelease version. The wide analyzer constraint combined with the somewhat out of date analyzer constraint in the released version is likely the reason.

You can run `pub upgrade -v` to see all the version solving in action, which is what I used for the analysis below.

It looks like it started the version solve from `aqueduct_test` and not `aqueduct`. At that point I think the solver essentially tries to get all the latest versions of the `aqueduct_test` dependencies. It will prefer non-dev releases but it will select them if it needs to. It does not do backtracking during this process and may end up selecting a dev release because of some earlier decision it made (such as choosing the most updated analyzer version).

I noticed initially that the `-b1` version allows a very permissive analyzer constraint, while the released version of aqueduct does not, and while I haven't fully confirmed I bet that is exactly what happened here (mostly a hunch, semi-backed up by some of the messages below).

It initially selects aqueduct 3.3.0 and starts trying to solve for its constraints, ultimately though it derives `not aqueduct >=3.3.0` with this message `fact: test >=1.3.0 <1.5.3-∞ or >=1.6.7 requires test_core 0.1.0 or 0.2.0 or 0.2.0+1 or 0.2.1 or matcher >=0.12.3 <0.12.4 or analyzer >=0.36.0 <0.40.0`.

Later on when trying a different path it ends up trying another combination of deps it ultimately works, but only allows the `3.3.0-b1` version with the below description. Note that it hasn't actually solved for analyzer yet this time (it started the solve with a different package as the root I think).

```
SLVR:                                                     fact: http_multi_server >=2.0.5 depends on async >=1.2.0 <3.0.0
SLVR:                                                       selecting http_multi_server 2.2.0
SLVR:                                                       fact: aqueduct >=3.2.1 <3.3.0-b1 depends on analyzer >=0.32.0 <0.36.0
SLVR:                                                       fact: aqueduct >=3.0.0 <3.3.0-b1 depends on safe_config ^2.0.0
SLVR:                                                       derived: not aqueduct >=3.2.1 <3.3.0-b1
SLVR:                                                       fact: aqueduct >=3.1.0 <3.2.1-∞ depends on analyzer >=0.32.0 <0.35.0
SLVR:                                                       derived: not aqueduct >=3.1.0 <3.2.1-∞
SLVR:                                                       fact: aqueduct >=3.0.0-beta.1 <3.1.0-beta.1 depends on analyzer ^0.32.0
SLVR:                                                       derived: not aqueduct >=3.0.0-beta.1 <3.1.0-beta.1
```

**Comment by jakemac53** [comment]
I do think it would be useful to have a flag which doesn't allow any prerelease versions at all (it probably should have been the default really).

**Comment by joeconwaystk** [comment]
Thank you for the breakdown, it was an interesting read. Agreed on the pre-release flag! I think that is a good solution and I wonder if adding it in as the default in future releases would be considering a breaking change? 

Separately, is there a way to remove version 3.2.1-b1 from pub.dev? I checked the new Admin console but I could only see transfer/deprecate.

**Comment by sigurdm** [comment]
> Separately, is there a way to remove version 3.2.1-b1 from pub.dev? I checked the new Admin console but I could only see transfer/deprecate.

No - not yet. We hope to build a feature where you can mark versions as broken: https://github.com/dart-lang/pub-dev/issues/378)

**Comment by sigurdm** [comment]
Closing this in favor of https://github.com/dart-lang/pub/issues/2471

**Comment by mit-mit** [comment]
@jakemac53 do you have a concrete suggestion for the semantics of the pre-release flag? If so, can you please comment over in #2471 ?
